rust/src/ike/detect.rs

```diff
@@ -0,0 +1,263 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+// Author: Frank Honza <frank.honza@dcso.de>
+
+use crate::ike::ike::*;
+use std::ptr;
+use std::ffi::CStr;
+use super::ipsec_parser::IkeV2Transform;
+
+#[no_mangle]
+pub extern "C" fn rs_ike_state_get_exch_type(
+    tx: &mut IKETransaction,
+    exch_type: *mut u8,
+) -> u8 {
+    if exch_type == std::ptr::null_mut() {
+        return 0;
+    }
+
+    if tx.ike_version == 1 {
+        if let Some(r) = tx.hdr.ikev1_header.exchange_type {
+            unsafe {
+                *exch_type = r;
+            }
+            return 1;
+        }
+    } else if tx.ike_version == 2 {
+        unsafe {
+            *exch_type = tx.hdr.ikev2_header.exch_type.0;
+        }
+        return 1;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_ike_state_get_spi_initiator(
+    tx: &mut IKETransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    if buffer == std::ptr::null_mut() || buffer_len == std::ptr::null_mut() {
+        return 0;
+    }
+
+    unsafe {
+        *buffer = tx.hdr.spi_initiator.as_ptr();
+        *buffer_len = tx.hdr.spi_initiator.len() as u32;
+    }
+    return 1;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_ike_state_get_spi_responder(
+    tx: &mut IKETransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    if buffer == std::ptr::null_mut() || buffer_len == std::ptr::null_mut() {
```

Comment:
I think we do not need these checks

Reply:
have added them everywhere after review from @chifflier ....  (this was in the first PR for the parser)

Reply:
@chifflier, why did you ask for this ?
These checks are not done in the other parsers, are they ? cf `rs_ssh_tx_get_protocol` for instance

Reply:
> If Suricata ensures that pointers are never null

I think that is the case now...
@jasonish thoughts on this ?

Reply:
I'm fine with it, even though I see it as part of the contract here that these should not be NULL.  Perhaps in the future we should start adding bug_on type assertions here to catch the possibility of these being null in CI.

Reply:
Yes `bug_on` or `debug_validate_bug_on` seem to be better than testing and returning 0 @fhonza